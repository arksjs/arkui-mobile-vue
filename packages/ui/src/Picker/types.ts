import type { PopupEmits, PopupProps, PopupRef } from '../popup/types'
import type { AnyObject, Noop } from '../helpers/types'
import type {
  SelectorValue,
  SelectorModelValue,
  SelectorDetail,
  SelectorValueParser,
  SelectorValueFormatter,
  SelectorOnChange
} from '../SelectorField/types'
import type { FocusWithoutEventEmits, FormItemCommonProps } from '../Form/types'

export interface UserFieldNames {
  label?: string
  value?: string
  children?: string
}
export interface FieldNames {
  label: string
  value: string
  children: string
}

export type UserOptionItem = string | number | AnyObject
export interface OptionItem {
  label: string
  value: string | number
  children: OptionItem[]
  disabled: boolean
}

export interface Col {
  key: string
  rows: ColRow[]
}

export interface ColRow {
  label: string
  value: string | number
  hasChildren: boolean
  indexes: number[]
  values?: SelectorValue[]
  disabled?: boolean
  selected?: boolean
}

export interface PickerOptionsHandler {
  (index: number, parent?: ColRow): ColRow[]
}
export interface DefaultValueGetter {
  (): SelectorValue[]
}
export interface PickerLabelFormatter {
  (labelArray: string[]): string
}
export interface PickerHandlers {
  optionsHandler?: PickerOptionsHandler
  defaultValueGetter?: DefaultValueGetter
  formatter: SelectorValueFormatter
  parser: SelectorValueParser
  labelFormatter: PickerLabelFormatter
}

export interface PickerDetail {
  value: (string | number)[]
  label: string
}

export type OnConfirm = (payload: PickerDetail) => void

/**
 * PickerCommon
 */
export interface PickerCommonProps {
  modelValue?: SelectorModelValue
  formatter?: SelectorValueFormatter
  parser?: SelectorValueParser
}

export interface PickerOptionsProps {
  options: UserOptionItem[] | UserOptionItem[][]
  fieldNames?: UserFieldNames
}

export interface PickerCommonEmits {
  onChange?: SelectorOnChange
}

export interface ShowPickerOptions extends PickerOptionsProps {
  title?: string
  value?: (string | number)[]
}

/**
 * PickerView
 */
export interface PickerViewProps
  extends PickerCommonProps,
    PickerOptionsProps {}

export type PickerViewEmits = PickerCommonEmits

export interface PickerViewRef {
  getDetail: () => SelectorDetail
  resize: Noop
}

/**
 * PickerPopup
 */
export interface PickerPopupProps
  extends PopupProps,
    PickerCommonProps,
    PickerOptionsProps {
  title?: string
}

export interface PickerPopupEmits extends PopupEmits, PickerCommonEmits {
  onConfirm?: OnConfirm
}

export interface PickerPopupRef extends PopupRef {
  getDetail: () => SelectorDetail
}

/**
 * Picker
 */
export interface PickerProps
  extends FormItemCommonProps,
    PickerCommonProps,
    PickerOptionsProps {
  placeholder?: string
}

export interface PickerEmits
  extends FocusWithoutEventEmits,
    PickerCommonEmits {}

export interface ScrollElement extends HTMLElement {
  scrolling?: boolean
  scrollTimer?: number
}

export type PickerViewAfterUpdate = (
  valueArray: readonly SelectorValue[],
  labelArray: readonly string[],
  cols: readonly Col[]
) => void

export type {
  OnConfirm as PickerOnConfirm,
  UserFieldNames as PickerFieldNames,
  SelectorOnChange as PickerOnChange
}
