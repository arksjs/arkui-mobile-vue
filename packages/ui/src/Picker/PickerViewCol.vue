<template>
  <div class="ak-picker-view_col">
    <VirtualList
      class="ak-picker-view_list"
      :data-index="listIndex"
      :ids="list.map(v => v.value)"
      :itemSize="getItemSize"
      @scroll.stop="onListScroll"
    >
      <template #default="{ index }">
        <div
          class="ak-picker-view_item"
          :class="{
            selected: list[index].selected,
            disabled: list[index].disabled
          }"
          :data-index="index"
        >
          {{ list[index].label }}
        </div>
      </template>
    </VirtualList>
  </div>
</template>

<script lang="ts">
import { defineComponent, PropType } from 'vue'
import { VirtualList } from '../VirtualList'
import type { ColRow } from './types'
import { DEFAULT_ITEM_HEIGHT } from './util'

export default defineComponent({
  name: 'ak-picker-view-col',
  components: { VirtualList },
  props: {
    list: {
      type: Array as PropType<ColRow[]>,
      required: true
    },
    listIndex: {
      type: Number,
      required: true
    },
    onListScroll: {
      type: Function as PropType<(payload: UIEvent) => void>,
      required: true
    }
  },
  setup(props) {
    function getItemSize(index: number) {
      if (props.list.length === 1) {
        return DEFAULT_ITEM_HEIGHT * 5
      }

      if (index === 0 || index >= props.list.length - 1) {
        return DEFAULT_ITEM_HEIGHT * 3
      }

      return DEFAULT_ITEM_HEIGHT
    }

    return {
      getItemSize
    }
  }
})
</script>
