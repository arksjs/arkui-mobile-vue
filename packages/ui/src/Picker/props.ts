import type { PropType } from 'vue'
import type { PropsToEmits } from '../helpers/types'
import type {
  PickerEmits,
  PickerPopupEmits,
  PickerCommonEmits,
  UserFieldNames,
  UserOptionItem
} from './types'
import type {
  SelectorValue,
  SelectorModelValue,
  SelectorDetail,
  SelectorValueParser,
  SelectorValueFormatter
} from '../SelectorField/types'
import { isStringNumberMix } from '../helpers/util'
import { getDefaultFieldNames } from './util'
import { formActiveEmits, formItemProps } from '../Form/form'
import { popupEmits, popupExtendProps } from '../popup/popup'

const isValue = (value: SelectorValue) => {
  return isStringNumberMix(value) || value instanceof Date
}

const isModelValue = (value: SelectorModelValue) => {
  if (Array.isArray(value)) {
    return (
      value.filter(v => {
        return !isValue(v)
      }).length === 0
    )
  }
  return isValue(value)
}

export const isPickerDetail = <T extends SelectorDetail = SelectorDetail>(
  detail: T
) => {
  return (
    detail && isModelValue(detail.value) && typeof detail.label === 'string'
  )
}

export const commonProps = {
  modelValue: {
    type: [Date, String, Number, Array] as PropType<SelectorModelValue>
  },
  options: {
    type: Array as PropType<UserOptionItem[]>,
    default: () => [] as UserOptionItem[]
  },
  fieldNames: {
    type: Object as PropType<UserFieldNames>,
    default: getDefaultFieldNames()
  },
  formatter: {
    type: Function as PropType<SelectorValueFormatter>
  },
  parser: {
    type: Function as PropType<SelectorValueParser>
  }
}

export const commonEmits: PropsToEmits<PickerCommonEmits> = {
  change: isModelValue,
  'update:modelValue': isModelValue
}

/**
 * view
 */
export const pickerViewProps = commonProps

export const pickerViewEmits = commonEmits

/**
 * popup
 */
export const pickerPopupProps = {
  ...popupExtendProps
}

export const pickerPopupEmits: PropsToEmits<PickerPopupEmits> = {
  ...popupEmits,
  ...commonEmits,
  confirm: (payload: SelectorDetail) => isPickerDetail(payload)
}

/**
 * input
 */
export const pickerProps = {
  ...formItemProps,
  placeholder: {
    type: String,
    default: ''
  }
}

export const pickerEmits: PropsToEmits<PickerEmits> = {
  ...commonEmits,
  ...formActiveEmits
}
