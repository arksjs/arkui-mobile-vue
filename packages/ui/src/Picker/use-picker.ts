import { nextTick, onMounted, ref, watch } from 'vue'
import type { SetupContext } from 'vue'
import { cloneData, isSameArray } from '../helpers/util'
import type {
  ColRow,
  OptionItem,
  PickerOptionsHandler,
  PickerHandlers,
  Col,
  PickerProps,
  PickerPopupEmits,
  PickerViewEmits,
  PickerViewProps,
  PickerPopupProps,
  PickerEmits,
  PickerViewRef
} from './types'
import type {
  SelectorValue,
  SelectorModelValue,
  SelectorDetail
} from '../SelectorField/types'
import {
  cloneDetail,
  isSameDetail,
  isSameValue,
  validateValues,
  getColRows,
  getFormatOptions,
  isValidValue
} from './util'
import type { PopupCustomConfirm } from '../popup/types'
import type { PropsToEmits } from '../helpers/types'

function getDefaultDetail(handlers: PickerHandlers) {
  return formatter([], [], handlers)
}

export function usePicker(
  props: Partial<PickerProps>,
  ctx: SetupContext<PropsToEmits<PickerEmits>>,
  {
    name,
    handlers
  }: {
    name: string
    handlers: PickerHandlers
  }
) {
  const root = ref<HTMLElement>()
  const { emit } = ctx
  const isInitPopup = ref(false)
  const popupVisible = ref(true)
  const fieldValue = ref('')
  const fieldLabel = ref('')
  const popup = ref()

  let detail = getDefaultDetail(handlers)

  function updateValue(val: unknown) {
    if (val == null) {
      // 解决 formily 强制null的问题
      return
    }

    if (!isValidValue(val)) {
      updateDetail(getDefaultDetail(handlers))
      return
    }

    if (!isSameValue(val, detail.value)) {
      // 如果value不同
      const { options, isCascade } = getFormatOptions(
        props.options || [],
        props.fieldNames || {},
        handlers.optionsHandler,
        name === 'cascader'
      )

      const { value, label, valid } = validateValues(
        parser(val, handlers),
        options,
        isCascade,
        handlers.optionsHandler
      )

      if (valid) {
        updateDetail(formatter(value, label, handlers))
      }
    }
  }

  function updateDetail(newDetail: SelectorDetail) {
    detail = newDetail

    fieldValue.value = detail.value != null ? detail.value.toString() : ''
    fieldLabel.value = detail.label
  }

  function onFieldClick() {
    if (!props.disabled) {
      if (!isInitPopup.value) {
        isInitPopup.value = true
      } else {
        popupVisible.value = true
      }
    }
  }

  function getDetail() {
    return cloneDetail(detail)
  }

  function onConfirm(newDetail: SelectorDetail) {
    if (!isSameDetail(newDetail, detail)) {
      updateDetail(newDetail)

      emit('update:modelValue', getDetail().value)
      emit('change', getDetail().value)
    }
  }

  watch(
    [() => props.modelValue, () => props.options],
    () => updateValue(props.modelValue),
    {
      deep: true,
      immediate: true
    }
  )

  watch([isInitPopup, popupVisible], ([isInit, visible]) => {
    if (isInit && visible) {
      emit('focus')
    } else if (!visible) {
      emit('blur')
    }
  })

  return {
    root,
    popup,
    isInitPopup,
    popupVisible,
    fieldValue,
    fieldLabel,
    onFieldClick,
    onConfirm
  }
}

export function usePickerPopup(
  props: Partial<PickerPopupProps>,
  { emit }: SetupContext<PropsToEmits<PickerPopupEmits>>,
  {
    customConfirm,
    onCancelClick
  }: {
    customConfirm: PopupCustomConfirm<SelectorDetail>
    onCancelClick: () => void
  },
  { handlers }: { handlers: PickerHandlers }
) {
  const viewRef = ref<PickerViewRef>()

  let detail = getDefaultDetail(handlers)

  function beforeConfirm() {
    const newDetail = getViewDetail()

    if (!isSameDetail(newDetail, detail)) {
      detail = newDetail

      // 跟picker-view不一样，改变数值时机是确定按钮
      emit('update:modelValue', getDetail().value)
      emit('change', getDetail().value)
    } else {
      detail = newDetail
    }

    return getDetail()
  }

  function onHeaderLeftClick() {
    onCancelClick()
  }

  function onHeaderRightClick() {
    customConfirm(beforeConfirm())
  }

  function getDetail() {
    return cloneDetail(detail)
  }

  function getViewDetail() {
    return viewRef.value?.getDetail() || getDefaultDetail(handlers)
  }

  watch(
    () => props.visible,
    val => {
      if (val && viewRef.value?.resize) {
        viewRef.value.resize()
      }
    }
  )

  watch(
    () => props.modelValue,
    val => {
      nextTick(() => {
        if (isValidValue(val)) {
          detail = getViewDetail()
        } else {
          // 主要是针对 pickerView 强制要求值得问题
          detail = getDefaultDetail(handlers)
        }
      })
    },
    {
      deep: true
    }
  )

  onMounted(() => {
    if (isValidValue(props.modelValue)) {
      detail = getViewDetail()
    }
  })

  return {
    viewRef,
    getDetail,
    onHeaderLeftClick,
    onHeaderRightClick
  }
}

interface ViewUseOptions {
  name: 'cascader' | 'picker'
  afterUpdate: (
    valueArray: SelectorValue[],
    labelArray: string[],
    cols: Col[]
  ) => void
  handlers: PickerHandlers
}

export function usePickerView(
  props: PickerViewProps,
  { emit }: SetupContext<PropsToEmits<PickerViewEmits>>,
  { name, afterUpdate, handlers }: ViewUseOptions
) {
  const cols = ref<Col[]>([])

  const options2 = ref<OptionItem[] | OptionItem[][]>([])
  const isCascade = ref(false)

  let selectedLabels: string[] = []
  let selectedValues: SelectorValue[] = []

  const currentLabels = ref<string[]>([])
  const currentValues = ref<SelectorValue[]>([])

  const isPicker = name === 'picker'

  const optionsHandler: PickerOptionsHandler | null =
    handlers.optionsHandler || null

  function updateOptions() {
    const { options, isCascade: isCascade2 } = getFormatOptions(
      props.options || [],
      props.fieldNames || {},
      optionsHandler,
      !isPicker
    )

    options2.value = options

    isCascade.value = isCascade2
  }

  function updateOriginalValue(val: SelectorValue[], forceUpdate = false) {
    const { valid, value: values } = validateValues(
      val,
      options2.value,
      isCascade.value,
      optionsHandler
    )

    if ((valid && !isSameArray(values, currentValues.value)) || forceUpdate) {
      update(values)

      currentLabels.value = values.length > 0 || isPicker ? selectedLabels : []
      currentValues.value = values.length > 0 || isPicker ? selectedValues : []
    }

    return getDetail()
  }

  function updateValue(val: unknown, forceUpdate = false) {
    return updateOriginalValue(parser(val, handlers), forceUpdate)
  }

  function getDetail() {
    return formatter(
      cloneData(currentValues.value),
      cloneData(currentLabels.value),
      handlers
    )
  }

  function addCache(item: { value: string | number; label: string }) {
    selectedValues.push(item.value)
    selectedLabels.push(item.label)
  }

  function update(selecteds: SelectorValue[]) {
    !isCascade.value ? updateCols(selecteds) : updateCascadeCols(selecteds)

    if (isPicker) {
      currentLabels.value = selectedLabels
      currentValues.value = selectedValues
    }

    afterUpdate(selectedValues, selectedLabels, cols.value)
  }

  function clearCache() {
    selectedLabels = []
    selectedValues = []
  }

  /**
   * 更新多列展示效果
   */
  function updateCols(selecteds: SelectorValue[]) {
    clearCache()

    const newCols: Col[] = []

    if (options2.value.length === 0) {
      cols.value = newCols
      return []
    }

    const options = (
      Array.isArray(options2.value[0]) ? options2.value : [options2.value]
    ) as OptionItem[][]

    options.forEach((subOptions, index) => {
      let hasSelected = false

      const rows: ColRow[] = []

      for (let i = 0; i < subOptions.length; i++) {
        const optionItem = subOptions[i]
        const row: ColRow = {
          value: optionItem.value,
          label: optionItem.label,
          hasChildren: false,
          indexes: [i],
          selected: false
        }

        if (optionItem.value == selecteds[index]) {
          hasSelected = true
          row.selected = true
          addCache(optionItem)
        }

        rows.push(row)
      }

      if (!hasSelected) {
        if (subOptions[0]) {
          rows[0].selected = true
          addCache(subOptions[0])
        }
      }

      newCols.push({
        key: index.toString(),
        rows
      })
    })
    cols.value = newCols

    return newCols
  }

  /**
   * 日期等更新模式
   * @param selecteds 选择值
   */
  function updateVirtualOptionsCols(selecteds: SelectorValue[]) {
    clearCache()

    if (selecteds.length === 0) {
      // 如果没有默认值，尝试获取默认值
      selecteds = getCascadeDefaultSelecteds()
    }

    const newCols: Col[] = []

    const getRows = optionsHandler as PickerOptionsHandler
    let colIndex = 0
    let rows = getRows(colIndex)
    let lastGroupSelected = false
    let key = ''
    let i = 0

    for (; i <= selecteds.length; i++) {
      const selected = selecteds[i]

      let nextParent: ColRow | null = null
      lastGroupSelected = false

      if (rows) {
        for (let j = 0; j < rows.length; j++) {
          const row = rows[j]

          if (selected != null && row.value === selected) {
            // 找到
            row.selected = true
            lastGroupSelected = true

            if (row.hasChildren) {
              nextParent = row
              colIndex++
            }

            addCache({
              label: row.label,
              value: row.value
            })
          }
        }

        newCols.push({
          key: `${i}-${key}`,
          rows
        })
      }

      if (!nextParent) {
        break
      } else {
        rows = getRows(colIndex, nextParent)
        key = nextParent.value.toString()
      }
    }

    if (isPicker && !lastGroupSelected) {
      const index = 0

      let lastColFirstRow: ColRow | null = rows[index]

      while (lastColFirstRow) {
        lastColFirstRow.selected = true
        addCache({
          label: lastColFirstRow.label,
          value: lastColFirstRow.value
        })
        i++

        if (lastColFirstRow.hasChildren) {
          colIndex++
          rows = getRows(colIndex, lastColFirstRow)
          newCols.push({
            key: `${i}-${lastColFirstRow.value.toString()}`,
            rows
          })

          lastColFirstRow = rows[0]
        } else {
          lastColFirstRow = null
        }
      }
    }
    cols.value = newCols

    return newCols
  }

  /**
   * 级联更新模式
   * @param selecteds 选择值
   */
  function updateCascadeCols(selecteds: SelectorValue[]) {
    if (typeof optionsHandler === 'function') {
      updateVirtualOptionsCols(selecteds)
      return
    }

    clearCache()

    const newCols: Col[] = []

    let optionList: OptionItem[] = options2.value as OptionItem[]

    if (optionList.length === 0) {
      cols.value = newCols
      return newCols
    }

    let rows = getColRows(optionList, [])

    if (selecteds.length === 0) {
      // 如果没有默认值，尝试获取默认值
      selecteds = getCascadeDefaultSelecteds()
    }

    let lastGroupSelected = false
    let key = ''
    let i = 0

    for (; i <= selecteds.length; i++) {
      const selected = selecteds[i]

      let nextParent: OptionItem | null = null
      let nextParentIndexes: number[] = []
      lastGroupSelected = false

      if (rows) {
        for (let j = 0; j < rows.length; j++) {
          const optionItem = optionList[j]
          const row = rows[j]

          if (selected != null && row.value === selected) {
            // 找到
            row.selected = true
            lastGroupSelected = true

            if (row.hasChildren) {
              nextParent = optionItem
              nextParentIndexes = row.indexes
            }

            addCache(optionItem)
          }
        }

        newCols.push({
          key: `${i}-${key}`,
          rows
        })
      }

      if (!nextParent) {
        break
      } else {
        optionList = nextParent.children
        rows = getColRows(optionList, [...nextParentIndexes])
        key = nextParent.value.toString()
      }
    }

    if (isPicker && !lastGroupSelected) {
      const index = 0

      let lastGroupFirstItem: OptionItem | null = optionList[index]

      while (lastGroupFirstItem) {
        const lastColFirstRow = rows[index]
        lastColFirstRow.selected = true
        addCache(lastGroupFirstItem)
        i++

        if (lastColFirstRow.hasChildren) {
          optionList = lastGroupFirstItem.children
          rows = getColRows(optionList, [...lastColFirstRow.indexes])
          newCols.push({
            key: `${i}-${lastColFirstRow.value.toString()}`,
            rows
          })

          lastGroupFirstItem = optionList[0]
        } else {
          lastGroupFirstItem = null
        }
      }
    }
    cols.value = newCols

    return newCols
  }

  /**
   * 非级联更新选择项
   */
  function updateColSelected(colIndex: number, current: number) {
    const newCol = cols.value[colIndex]

    newCol &&
      newCol.rows.forEach((row, index) => {
        if (index === current) {
          currentValues.value[colIndex] = row.value
          currentLabels.value[colIndex] = row.label
          row.selected = true
        } else {
          row.selected = false
        }
      })
  }

  /**
   * 获取默认选择数据
   * @summary 主要用于一些日期啥的，可以默认当天
   */
  function getCascadeDefaultSelecteds() {
    const selecteds = handlers.defaultValueGetter
      ? handlers.defaultValueGetter()
      : []

    if (selecteds.length > 0) {
      return selecteds
    }

    function getFirstSelected(
      values: SelectorValue[],
      optionList: OptionItem[]
    ): SelectorValue[] {
      const optionItem = optionList[0]

      if (optionItem) {
        values.push(optionItem.value)

        if (optionItem.children.length > 0) {
          return getFirstSelected(values, optionItem.children)
        }
      }

      return values
    }

    return !isPicker ? [] : getFirstSelected([], options2.value as OptionItem[])
  }

  function getValuesByRow(row: ColRow) {
    if (row.values) {
      return row.values as SelectorValue[]
    }

    const values: SelectorValue[] = []
    const indexes = row.indexes
    let i = 0
    let options = options2.value as OptionItem[]
    let optionItem = options[indexes[i]]

    while (options.length > 0 && i < indexes.length && optionItem) {
      values.push(optionItem.value)
      options = optionItem.children
      i++
      optionItem = options[indexes[i]]
    }

    return values
  }

  function onChange() {
    emit('update:modelValue', getDetail().value)
    emit('change', getDetail().value)
  }

  watch(
    [() => props.options, () => props.fieldNames],
    () => {
      updateOptions()
      updateOriginalValue(currentValues.value, true)
    },
    {
      deep: true,
      immediate: true
    }
  )

  watch(
    () => props.modelValue,
    val => updateValue(val),
    {
      deep: true,
      immediate: true
    }
  )

  // picker 要默认数据
  if (
    isPicker &&
    (!isValidValue(props.modelValue) ||
      !isSameValue(props.modelValue, currentValues.value))
  ) {
    // 如果传入的数据
    onChange()
  }

  return {
    cols,
    currentLabels,
    currentValues,
    isCascade,
    getDetail,
    update,
    updateColSelected,
    getValuesByRow,
    updateOriginalValue,
    onChange
  }
}

type PickerFormatter = (
  valueArray: SelectorValue[],
  labelArray: string[],
  handlers: PickerHandlers
) => SelectorDetail

type PickerParser = (
  value: unknown,
  handlers: PickerHandlers
) => SelectorValue[]

const formatter: PickerFormatter = (valueArray, labelArray, handlers) => {
  const defaultLabel = handlers.labelFormatter(labelArray)
  const ret = handlers.formatter(valueArray, labelArray)

  if ((ret as SelectorDetail)?.value != null) {
    return {
      value: (ret as SelectorDetail).value,
      label: (ret as SelectorDetail).label ?? ''
    }
  }

  return {
    value: ret as SelectorModelValue,
    label: defaultLabel
  }
}

const parser: PickerParser = (val, handlers) => {
  return handlers.parser(val)
}
