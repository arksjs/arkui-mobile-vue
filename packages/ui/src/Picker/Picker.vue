<template>
  <div class="fx-picker" :class="{ disabled }" ref="root">
    <SelectorField
      :label="fieldLabel"
      :value="fieldValue"
      :disabled="disabled"
      :name="name"
      :placeholder="placeholder"
      @fieldClick="onFieldClick"
    />
    <PickerPopup
      :options="options"
      :fieldNames="fieldNames"
      :modelValue="modelValue"
      :title="placeholder"
      :formatter="formatter"
      :parser="parser"
      v-model:visible="popupVisible"
      v-if="isInitPopup"
      @confirm="onConfirm"
      ref="popup"
    />
  </div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'
import { SelectorField } from '../SelectorField'
import PickerPopup from './PickerPopup.vue'
import { pickerEmits, pickerProps, commonProps } from './props'
import { mergeHandlers } from './util'
import { usePicker } from './use-picker'

export default defineComponent({
  name: 'fx-picker',
  components: { SelectorField, PickerPopup },
  props: { ...commonProps, ...pickerProps },
  emits: { ...pickerEmits },
  setup(props, ctx) {
    return {
      ...usePicker(props, ctx, {
        name: 'picker',
        handlers: mergeHandlers({
          formatter: props.formatter,
          parser: props.parser
        })
      })
    }
  }
})
</script>
