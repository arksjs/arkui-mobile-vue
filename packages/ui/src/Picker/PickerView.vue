<template>
  <div class="ta-picker-view" ref="root">
    <div class="ta-picker-view_cols">
      <ViewCol
        v-for="(colItem, listIndex) in cols"
        :key="colItem.key"
        :list="colItem.rows"
        :listIndex="listIndex"
        :onListScroll="onListScroll"
      />
      <Empty :description="locale.pickerEmptyText" v-if="cols.length === 0" />
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent, nextTick, inject, shallowRef } from 'vue'
import { Empty } from '../Empty'
import { frameTo, type PropsToEmits } from '../helpers'
import { pickerViewEmits, pickerViewProps } from './props'
import { usePickerView } from '../Picker/use-picker'
import type { PickerHandlers, PickerViewEmits } from './types'
import { useLocale } from '../ConfigProvider/context'
import ViewCol from './PickerViewCol.vue'
import { DEFAULT_ITEM_HEIGHT, mergeHandlers } from './util'

interface ScrollElement extends HTMLElement {
  scrolling?: boolean
  scrollTimer?: number
}

export default defineComponent({
  name: 'ta-picker-view',
  components: { Empty, ViewCol },
  props: {
    ...pickerViewProps
  },
  emits: { ...pickerViewEmits } as PropsToEmits<PickerViewEmits>,
  setup(props, ctx) {
    const { locale } = useLocale()
    const root = shallowRef<HTMLElement | null>(null)
    const handlers = inject<Partial<PickerHandlers>>('taPickerHandlers', {})

    const { getDetail, cols, isCascade, update, updateColSelected, getValuesByRow, onChange } =
      usePickerView(props, ctx, {
        name: 'picker',
        afterUpdate: updatePos,
        handlers: mergeHandlers(
          {
            formatter: props.formatter,
            parser: props.parser
          },
          handlers
        )
      })

    function updatePos() {
      nextTick(() => {
        // 把选择数据展示在选择框内
        const $picker = root.value as HTMLElement

        if ($picker) {
          const $lists = $picker.querySelectorAll(`.ta-picker-view_list`)

          $lists.forEach(($el, index) => {
            const $list = $el as ScrollElement

            if (cols.value[index] && $list.scrolling !== true) {
              const list = cols.value[index]
              let selectIndex = -1

              for (let i = 0, len = list.rows.length; i < len; i++) {
                if (list.rows[i].selected) {
                  selectIndex = i
                  break
                }
              }

              requestAnimationFrame(() => {
                $list.scrollTop = DEFAULT_ITEM_HEIGHT * selectIndex
              })
            }
          })
        }
      })
    }

    function onListScroll(e: Event) {
      const $list = e.currentTarget as ScrollElement

      if ($list.scrolling) {
        return
      }

      clearTimeout($list.scrollTimer)

      const itemHeight = DEFAULT_ITEM_HEIGHT
      const colIndex = parseInt($list.dataset.index as string)
      const list = cols.value[colIndex]
      let current = Math.round($list.scrollTop / itemHeight)
      let oldSelectIndex = 0

      for (let i = 0; i < list.rows.length; i++) {
        if (list.rows[i].selected) {
          oldSelectIndex = i
          break
        }
      }

      let isChange = current !== oldSelectIndex
      if (isChange) {
        while (list.rows[current].disabled && current !== oldSelectIndex) {
          // 处理disabled 不能选的问题
          if (current > oldSelectIndex) {
            current--
          } else {
            current++
          }
        }
      }
      isChange = current !== oldSelectIndex

      const item = list.rows[current]

      if (current * itemHeight === $list.scrollTop && !isChange) {
        // 如果一致 就不需要修正了
      } else {
        $list.scrollTimer = window.setTimeout(() => {
          // $list.scrollTop = current * itemHeight
          $list.scrolling = true

          frameTo({
            from: $list.scrollTop,
            to: current * itemHeight,
            duration: 100,
            progress(res) {
              $list.scrollTop = res.current
            },
            complete() {
              $list.scrolling = false
            }
          })

          if (isChange) {
            if (isCascade.value) {
              update(getValuesByRow(item))
            } else {
              updateColSelected(colIndex, current)
            }

            onChange()
          }
        }, 400)
      }
    }

    return {
      root,
      cols,
      getDetail,
      onListScroll,
      resize: updatePos,
      locale
    }
  }
})
</script>
