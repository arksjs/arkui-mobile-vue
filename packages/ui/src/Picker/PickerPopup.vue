<template>
  <Drawer
    class="fx-picker-popup"
    placement="bottom"
    :visible="visible"
    @visibleStateChange="onVisibleStateChange"
    @cancel="onCancel"
    @confirm="onConfirm"
    @update:visible="onUpdateVisible"
    ref="popup"
  >
    <template #header>
      <NavBar
        class="fx-drawer_header"
        :title="title"
        :leftButtons="[{ text: locale.pickerCancelText, type: 'primary' }]"
        :rightButtons="[{ text: locale.pickerConfirmText, type: 'primary' }]"
        :iconOnly="false"
        @leftButtonClick="onHeaderLeftClick"
        @rightButtonClick="onHeaderRightClick"
      >
      </NavBar>
    </template>
    <PickerView
      ref="viewRef"
      :modelValue="modelValue"
      :options="options"
      :fieldNames="fieldNames"
      :formatter="formatter"
      :parser="parser"
    />
  </Drawer>
</template>

<script lang="ts">
import { defineComponent, inject } from 'vue'
import PickerView from './PickerView.vue'
import { Drawer } from '../Drawer'
import { NavBar } from '../NavBar'
import { usePopupExtend } from '../popup/use-popup'
import { pickerPopupProps, pickerPopupEmits, commonProps } from './props'
import { usePickerPopup } from '../Picker/use-picker'
import { useLocale } from '../ConfigProvider/context'
import type { SelectorDetail } from '../SelectorField/types'
import type { PickerHandlers } from './types'
import { mergeHandlers } from './util'

export default defineComponent({
  name: 'fx-picker-popup',
  components: { PickerView, Drawer, NavBar },
  props: {
    ...pickerPopupProps,
    ...commonProps,
    title: {
      type: String,
      default: ''
    }
  },
  emits: {
    ...pickerPopupEmits
  },
  setup(props, ctx) {
    const { locale } = useLocale()
    const handlers = inject<Partial<PickerHandlers>>('fxPickerHandlers', {})

    const popup = usePopupExtend<SelectorDetail>(ctx)
    const pickerPopup = usePickerPopup(props, ctx, popup, {
      handlers: mergeHandlers(
        {
          formatter: props.formatter,
          parser: props.parser
        },
        handlers
      )
    })

    return {
      ...popup,
      ...pickerPopup,
      locale
    }
  }
})
</script>
