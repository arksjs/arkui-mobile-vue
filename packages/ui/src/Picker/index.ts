import Picker from './Picker.vue'
import PickerPopup from './PickerPopup.vue'
import PickerView from './PickerView.vue'
import { createConfirmHook, createShowPopup } from '../popup/api'
import type { PopupSuccessConfirmArgs } from '../popup/types'
import type { ShowPickerOptions, PickerDetail } from './types'

const showPicker = createShowPopup<ShowPickerOptions, PopupSuccessConfirmArgs<PickerDetail>>({
  apiName: 'showPicker',
  component: PickerPopup,
  createHook: createConfirmHook
})

export { Picker, PickerPopup, PickerView, showPicker }
export default Picker
