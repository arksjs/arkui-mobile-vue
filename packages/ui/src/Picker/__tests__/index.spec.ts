import { describe, test, expect } from 'vitest'
import { shallowMount } from '@vue/test-utils'
import { Picker, PickerPopup, PickerView } from '@/Picker'

describe('Picker', () => {
  test('snapshot', () => {
    const wrapper = shallowMount(Picker, {
      props: {
        options: []
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})

describe('PickerPopup', () => {
  test('snapshot', () => {
    const wrapper = shallowMount(PickerPopup, {
      props: {
        options: []
      },
      global: {
        stubs: { teleport: true }
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})

describe('PickerView', () => {
  test('snapshot', () => {
    const wrapper = shallowMount(PickerView, {
      props: {
        options: []
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
