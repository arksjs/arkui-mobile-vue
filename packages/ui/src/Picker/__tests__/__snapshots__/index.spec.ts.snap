// Vitest Snapshot v1

exports[`Picker > snapshot 1`] = `
"<div class=\\"ta-picker\\">
  <selector-field-stub label=\\"\\" name=\\"\\" placeholder=\\"\\" disabled=\\"false\\" value=\\"\\"></selector-field-stub>
  <!--v-if-->
</div>"
`;

exports[`PickerPopup > snapshot 1`] = `"<drawer-stub visible=\\"false\\" maskclosable=\\"true\\" placement=\\"bottom\\" showclose=\\"false\\" enablesafeareainsets=\\"true\\" showmask=\\"true\\" initialenableblurcancel=\\"true\\" class=\\"ta-picker-popup\\"></drawer-stub>"`;

exports[`PickerView > snapshot 1`] = `
"<div class=\\"ta-picker-view\\">
  <div class=\\"ta-picker-view_cols\\">
    <empty-stub description=\\"暂无选择可选\\" type=\\"default\\"></empty-stub>
  </div>
</div>"
`;
